### ISSUE TITLE
i'm confusing about p_ptr variable in ptrMemDecoder function in decoder?

### ISSUE THREAD

**Issue by minhcd1997** [HLTCHKUST/Mem2Seq#20]
` def ptrMemDecoder(self, enc_query, last_hidden):
        embed_q = self.C[0](enc_query) # b * e
        output, hidden = self.gru(embed_q.unsqueeze(0), last_hidden)
        temp = []
        u = [hidden[0].squeeze()]   
        for hop in range(self.max_hops):
            m_A = self.m_story[hop]
            if(len(list(u[-1].size()))==1): u[-1] = u[-1].unsqueeze(0) ## used for bsz = 1.
            u_temp = u[-1].unsqueeze(1).expand_as(m_A)
            prob_lg = torch.sum(m_A*u_temp, 2)
            prob_   = self.softmax(prob_lg)
            m_C = self.m_story[hop+1]
            temp.append(prob_)
            prob = prob_.unsqueeze(2).expand_as(m_C)
            o_k  = torch.sum(m_C*prob, 1)
            if (hop==0):
                p_vocab = self.W1(torch.cat((u[0], o_k),1))
            u_k = u[-1] + o_k
            u.append(u_k)
        p_ptr = prob_lg 
        return p_ptr, p_vocab, hidden
`
In the function ptrMemDecoder in the decoder, i have a question why `p_ptr = prob_lg` but not `p_ptr = self.softmax(prob_lg)`?
i have read in the paper, and it said that: 
>  P(ptr) is generated using the attention weights at the last MemNN hop of the decoder: P(ptr) = p(K)(t)

**Comment by andreamad8** [comment]
Hi @minhcd1997,

because we need the logit for the pointer supervision. Indeed we use CrossEntropyLoss which already has Softmax inside. Notice that prob_lg is the last hop logit and the softmax is just a normalization needed for the loss. Indeed when we greedy decode we do not need to use softmax. 

I hope this help, let me know if you have any other doubt.

Best Regards

Andrea

**Comment by minhcd1997** [comment]
thanks for your explanation, this helps me a lot
